### PR TITLE
 Support building against llvm15

### PR DESCRIPTION
**Pull request by yashssh**
Add support for llvmlite to be built against llvm15 while keeping support for llvm14 as well.

    1) Drop 2 passes no longer supported by old PM
    2) Update domTree passes to their wrapper passes
    3) Update getAddress() api
    4) Disable opaque pointer mode
    5) Give warning when build with llvm15
    
All the changes are put inside llvm version check guard so that it builds with both llvm14 and llvm15

### CODE REVIEW

**Comment by yashssh** [comment]
Removing `draft` tag as it won't allow me to add reviewers. Adding "draft" to the title instead.

File: ffi/Makefile.linux
Comment (yashssh) [inline ffi/Makefile.linux:5]: Kindly ignore this. I was facing some difficulties building locally so had to turn LTO off until I sort that.

Comment (gmarkall) [reply ffi/Makefile.linux:5]:
```suggestion
CXX_FLTO_FLAGS ?= -flto
```

File: ffi/core.cpp
Comment (gmarkall) [inline ffi/core.cpp:27]: Note to other reviewers - this does result in opaque pointers being disabled for the global context multiple times, but this operation is idempotent.

File: ffi/core.h
Comment (gmarkall) [inline ffi/core.h:7]: In particular, this defines the `LLVM_VERSION_MAJOR` macro referenced in various places in this PR.

File: llvmlite/tests/test_binding.py
Comment (gmarkall) [inline llvmlite/tests/test_binding.py:657]: LLVM 14 defaults to the soft-float ABI, LLVM 15 defaults to the hard-float ABI. We explicitly specify the soft-float ABI for consistency with the expected output. Note that hard float and hard double ABIs are tested in the subsequent two tests below.

File: llvmlite/tests/test_binding.py
Comment (gmarkall) [inline llvmlite/tests/test_binding.py:787]: I think this is a leftover from prior to when you had the test passing.
```suggestion
    @unittest.skip('Edit this test to test for both llvm14 and llvm15')
```

File: llvmlite/tests/test_binding.py
Comment (gmarkall) [inline llvmlite/tests/test_binding.py:792]: (Max patch versions are 14.0.6 and 15.0.7)

**Review by gmarkall — APPROVED** [review]
The code changes here have my approval - I've already done a couple of rounds of review internally on this PR, and I made a few edits for a couple of left over changes that needed removing (adjustments to the Makefile, and a test that was accidentally skipped).

I added the CI config so we can test with LLVM 15 from conda-forge on each platform - someone else might want to review those changes.

I ran the Numba test suite with this PR locally, and didn't encounter any issues - the two passes that aren't available with the legacy pass manager anymore (arg promotion and loop unswitch) don't seem to be used by Numba.

As I see it, this PR provides us a way to give experimental support for LLVM 15 once merged, with no further blockers. Following that, I think we can:

- Build an LLVM 15 llvmdev package and move testing to LLVM 15 exclusively, with the aim of officially moving to 15.
- Move to the new pass manager, and re-enable the passes that have to be disabled with LLVM 15 at present.
- Switch to opaque pointers (at least in the binding layer)
- Move to LLVM 16+

Some discussion of the ordering / dependencies of these is warranted though - perhaps at the next office hours.

File: ffi/passmanagers.cpp
Comment (modiking) [inline ffi/passmanagers.cpp:164]: Suggest being consistent with the #ifs, either all "<= 14" or all "< 15"

File: ffi/passmanagers.cpp
Comment (modiking) [inline ffi/passmanagers.cpp:350]: For reference the loop unswitch pass has been renamed to `SimpleLoopUnswitchPass` in the new pass manager (https://reviews.llvm.org/D124376). Naturally to enable this we need to move to NPM so it makes sense that this is disabled for now but it can come back with NPM.

File: ffi/passmanagers.cpp
Comment (modiking) [inline ffi/passmanagers.cpp:259]: Ditto with `ArgumentPromotionPass` in NPM as below

File: ffi/passmanagers.cpp
Comment (modiking) [reply ffi/passmanagers.cpp:350]: Actually `createSimpleLoopUnswitchLegacyPass` still exists in LLVM-15 so it can be dropped in here without issue.

**Review by modiking — COMMENTED** [review]
Thanks for putting this up!

File: ffi/passmanagers.cpp
Comment (modiking) [reply ffi/passmanagers.cpp:350]:
```
API_EXPORT(void)
LLVMPY_AddLoopUnswitchPass(LLVMPassManagerRef PM, bool optimizeForSize,
                           bool hasBranchDivergence) {
#if LLVM_VERSION_MAJOR < 15
    unwrap(PM)->add(
        createLoopUnswitchPass(optimizeForSize, hasBranchDivergence));
#else
    unwrap(PM)->add(
        createSimpleLoopUnswitchLegacyPass(optimizeForSize));
#endif
}
```

Comment (yashssh) [reply ffi/passmanagers.cpp:350]: Thanks for finding this, I have updated the code.

File: ffi/passmanagers.cpp
Comment (yashssh) [reply ffi/passmanagers.cpp:260]: Did you mean legacy version of this pass is available as well? I couldn't find it.

**Review by gmarkall — APPROVED** [review]
Re-approving following latest changes.

**Comment by modiking** [comment]
LGTM

**Comment by gmarkall** [comment]
Added "Pending Buildfarm" to make sure this is all OK on all other platforms.

**Comment by sklam** [comment]
BFID: `llvmlite_275`

**Comment by sklam** [comment]
`llvmlite_275` passed

**Comment by sklam** [comment]
Also passed buildfarm test against numba main (`numba_yaml_448`)

**Review by sklam — APPROVED** [review]

**Pull request merged**
